### PR TITLE
[GlobalISel][AArch64] AArch64O0PreLegalizerCombiner: Disable fixed-point iteration

### PR DESCRIPTION
**Pull request by tobias-stadler**
This adds an option to CombinerInfo to turn off the fixed-point iteration in the Combiner. This option is then used for AArch64O0PreLegalizerCombiner. The combines there are simple enough that code quality impact should be minimal with the current heuristics (instructions are processed from top to bottom of the basic block, new/changed instructions are added back to the worklist). Test changes are due to some instructions not being DCE'd, which has no actual impact because InstructionSelect performs DCE as well.

AArch64 CTMark O0:
-0.9% geomean compile-time (instruction count), no regressions
no change in size..text for any of the benchmarks

### CODE REVIEW

**Comment by tschuett** [comment]
For -O0, we use a special group of combines:
https://github.com/llvm/llvm-project/blob/e325e2e5928d4893bedee2d4bbb7d71cc855c67a/llvm/include/llvm/Target/GlobalISel/Combine.td#L1813

Is the one round vs fixed-point iteration issue limited to the -O0 combiner or is it a general issue of our combiners?

One advertisement of the Gisel combiner was that it could write a JSON file to the disk after a run. It would contain insights about which combines hit, which are nice to have, and compile time statistics.

**Comment by tobias-stadler** [comment]
> Is the one round vs fixed-point iteration issue limited to the -O0 combiner or is it a general issue of our combiners?

In my opinion, it's a general issue. The fixed-point iteration unnecessarily burns a lot of compile-time in every combiner. Turning fixed-point iteration off for O0 is inconsequential according to CTMark and the optnone_combines are very simple, so I don't expect a lot of regressions in other benchmarks. >O0 requires a more elaborate discussion. Turning it off for AArch64PreLegalizerCombiner and AArch64PostLegalizerCombiner and running CTMark O2, I get:
```
Program                                       compile_instructions                        size..text
                                              base-O2              patch2-O2       diff   base-O2    patch2-O2 diff
7zip/7zip-benchmark                           207743211267.00      206832115055.00 -0.44% 824480.00  824636.00 0.02%
Bullet/bullet                                 100680312509.00      100194629825.00 -0.48% 517188.00  517440.00 0.05%
kimwitu++/kc                                   42193871512.00       41963559810.00 -0.55% 461696.00  463252.00 0.34%
SPASS/SPASS                                    44102375151.00       43822705691.00 -0.63% 443008.00  443008.00 0.00%
tramp3d-v4/tramp3d-v4                          67051760706.00       66593264326.00 -0.68% 570940.00  571004.00 0.01%
ClamAV/clamscan                                54709178376.00       54184088001.00 -0.96% 456736.00  456752.00 0.00%
mafft/pairlocalalign                           33428526057.00       33104690017.00 -0.97% 321664.00  321664.00 0.00%
lencod/lencod                                  58348756630.00       57763311078.00 -1.00% 546524.00  546828.00 0.06%
consumer-typeset/consumer-typeset              35813224875.00       35401611592.00 -1.15% 419612.00  419632.00 0.00%
sqlite3/sqlite3                                35943993743.00       35457465790.00 -1.35% 436272.00  436272.00 0.00%
                           Geomean difference                                      -0.82%                      0.05%
```
All the code-size regressions originate from AArch64PreLegalizerCombiner and I still need to investigate them further. My hope is that a small additional heuristic in Combiner::WorklistMaintainer will fix most of the regressions, but my is guess that there will always be missed optimizations without fixed-point iteration for more complex combines, so the argument for disabling fixed-point iteration for O2 is not as clear-cut as for O0. Maybe O1?

**Comment by aemerson** [comment]
How are you collecting the compile_instructions metric?

**Comment by tobias-stadler** [comment]
> Instead of the boolean `EnableFixedPointIteration`, I would prefer an upper limit on the number of iterations to have more control.
I don't see where this would be needed. The overwhelming majority of work is done in the first iteration. Limiting the number of iterations to 2 already performs identically to fixed-point iteration on CTMark O2.

**Comment by tobias-stadler** [comment]
> How are you collecting the compile_instructions metric?

I run the test-suite on Linux under an AArch64 cross-toolchain (RelWithDebInfo clang, official ARM GNU linker/libc binaries) using `-DTEST_SUITE_USE_PERF=1` , which collects the instruction count and dumps it to a file (together with wall-time). I couldn't find a way to aggregate the instruction count, so I have slightly modified the lit module that collects the compile_time: https://github.com/tobias-stadler/llvm-test-suite/commit/8d68974d60831d4a2c2180b28a6dd0e9d6f9f192
I want to file a PR for this after some more cleanup.
I do this, because instruction count is much more stable than wall-time (<+-0.1% for a single run) and also is the metric that is measured on llvm-compile-time-tracker.

**Comment by arsenm** [comment]
> 
> I don't see where this would be needed. The overwhelming majority of work is done in the first iteration. Limiting the number of iterations to 2 already performs identically to fixed-point iteration on CTMark O2.
> 
> 
> 
> 


This would better match how instcombine handles this

File: llvm/lib/CodeGen/GlobalISel/Combiner.cpp [hunk 1]
```diff
@@ -27,6 +28,12 @@
 
 using namespace llvm;
 
+STATISTIC(NumOneIteration, "Number of functions with one iteration");
+STATISTIC(NumTwoIterations, "Number of functions with two iterations");
+STATISTIC(NumThreeIterations, "Number of functions with three iterations");
+STATISTIC(NumFourOrMoreIterations,
+          "Number of functions with four or more iterations");
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/Combiner.cpp:35]: Not sure all this variety is needed? We can probably prune this from instcombine now

**Review by arsenm — COMMENTED** [review]
lgtm but I think we can do with fewer statistics

File: llvm/lib/Target/AArch64/GISel/AArch64O0PreLegalizerCombiner.cpp [hunk 2]
```diff
@@ -165,6 +165,8 @@ bool AArch64O0PreLegalizerCombiner::runOnMachineFunction(MachineFunction &MF) {
   CombinerInfo CInfo(/*AllowIllegalOps*/ true, /*ShouldLegalizeIllegal*/ false,
                      /*LegalizerInfo*/ nullptr, /*EnableOpt*/ false,
                      F.hasOptSize(), F.hasMinSize());
+  CInfo.MaxIterations = 1;
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/GISel/AArch64O0PreLegalizerCombiner.cpp:168]: Please add a comment, perhaps calling back to this PR # so readers will know why we set it to 1.

File: llvm/lib/CodeGen/GlobalISel/Combiner.cpp (hunk 1 above)
Comment (tobias-stadler) [reply llvm/lib/CodeGen/GlobalISel/Combiner.cpp]: One, Two and Three are hit in CTMark. I have removed Four.

**Review by aemerson — APPROVED** [review]

**Pull request merged**